### PR TITLE
Adds a home method to device classes.

### PR DESCRIPTION
**Pull request by godofredoc**
This is necessary to ensure all the android tests start from the home
screen. Some tests like the android views crash if they are started with
the top level settings menu visible.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: dev/devicelab/lib/framework/devices.dart [hunk 1]
```diff
@@ -127,6 +126,9 @@ abstract class Device {
   /// Emulates pressing the power button, toggling the device's on/off state.
   Future<void> togglePower();
 
+  /// Emulates pressing the home button.
+  Future<void> home();
```
Comment (godofredoc) [inline dev/devicelab/lib/framework/devices.dart:130]: This the new method added to the interface.

File: dev/devicelab/lib/framework/devices.dart [hunk 2]
```diff
@@ -697,6 +692,11 @@ class AndroidDevice extends Device {
   Future<void> reboot() {
     return adb(<String>['reboot']);
   }
+
+  @override
+  Future<void> home() async {
```
Comment (godofredoc) [inline dev/devicelab/lib/framework/devices.dart:697]: This is the new interface method implementation for android.

File: dev/devicelab/lib/framework/devices.dart [hunk 3]
```diff
@@ -923,11 +917,16 @@ class IosDevice extends Device {
   Future<void> reboot() {
     return Process.run('idevicediagnostics', <String>['restart', '-u', deviceId]);
   }
+
+  @override
+  Future<void> home() async {
```
Comment (godofredoc) [inline dev/devicelab/lib/framework/devices.dart:922]: Noop for ios, fuchsia and fakedevice.

File: dev/devicelab/lib/framework/framework.dart [hunk 4]
```diff
@@ -144,6 +148,8 @@ class _TaskRunner {
       }
 
       final Device? device = await _getWorkingDeviceIfAvailable();
+      // Some tests assume the phone is in home
+      await device?.home();
```
Comment (godofredoc) [inline dev/devicelab/lib/framework/framework.dart:152]: This is the single use of the new method.

**Review by godofredoc — COMMENTED** [review]
Most of the changes are related to formatting. Added comments to the new added functionality.

File: dev/devicelab/lib/framework/devices.dart [hunk 5]
```diff
@@ -364,9 +365,8 @@ class FuchsiaDeviceDiscovery implements DeviceDiscovery {
   /// Picks the first connected Fuchsia device.
   @override
   Future<void> chooseWorkingDevice() async {
-    final List<FuchsiaDevice> allDevices = (await discoverDevices())
-      .map<FuchsiaDevice>((String id) => FuchsiaDevice(deviceId: id))
-      .toList();
+    final List<FuchsiaDevice> allDevices =
+        (await discoverDevices()).map<FuchsiaDevice>((String id) => FuchsiaDevice(deviceId: id)).toList();
```
Comment (dnfield) [inline dev/devicelab/lib/framework/devices.dart:369]: This line doesn't match the style guide's advice about keeping things on the same line. Previous formatting was more readable IMHO

File: dev/devicelab/lib/framework/devices.dart [hunk 6]
```diff
@@ -534,21 +533,23 @@ class AndroidDevice extends Device {
   }
 
   /// Executes [command] on `adb shell` and returns its exit code.
-  Future<void> shellExec(String command, List<String> arguments, { Map<String, String>? environment, bool silent = false }) async {
+  Future<void> shellExec(String command, List<String> arguments,
+      {Map<String, String>? environment, bool silent = false}) async {
```
Comment (dnfield) [inline dev/devicelab/lib/framework/devices.dart:537]: Maybe another trailing comma after `silent = false`, and below?

File: dev/devicelab/lib/framework/devices.dart [hunk 7]
```diff
@@ -728,12 +728,12 @@ class IosDeviceDiscovery implements DeviceDiscovery {
   /// [workingDevice].
   @override
   Future<void> chooseWorkingDevice() async {
-    final List<IosDevice> allDevices = (await discoverDevices())
-      .map<IosDevice>((String id) => IosDevice(deviceId: id))
-      .toList();
+    final List<IosDevice> allDevices =
+        (await discoverDevices()).map<IosDevice>((String id) => IosDevice(deviceId: id)).toList();
```
Comment (dnfield) [inline dev/devicelab/lib/framework/devices.dart:732]: less readable this way.

File: dev/devicelab/lib/framework/devices.dart [hunk 8]
```diff
@@ -852,16 +850,12 @@ class IosDevice extends Device {
...
-      .transform<String>(const Utf8Decoder(allowMalformed: true))
-      .listen((String line) {
-        sink.write(line);
-      });
-    _loggingProcess!.stderr
-      .transform<String>(const Utf8Decoder(allowMalformed: true))
-      .listen((String line) {
-        sink.write(line);
-      });
+    _loggingProcess!.stdout.transform<String>(const Utf8Decoder(allowMalformed: true)).listen((String line) {
```
Comment (dnfield) [inline dev/devicelab/lib/framework/devices.dart:853]: ditto

File: dev/devicelab/lib/tasks/perf_tests.dart [hunk 9]
```diff
@@ -740,7 +735,8 @@ class DevtoolsStartupTest {
       unawaited(process.exitCode.whenComplete(() {
         didExit = true;
       }));
-      await Future.any(<Future<void>>[completer.future, Future<void>.delayed(const Duration(minutes: 5)), process.exitCode]);
+      await Future.any(
+          <Future<void>>[completer.future, Future<void>.delayed(const Duration(minutes: 5)), process.exitCode]);
```
Comment (dnfield) [inline dev/devicelab/lib/tasks/perf_tests.dart:739]: Ditto.

File: dev/devicelab/lib/framework/devices.dart (hunk 5 above)
Comment (godofredoc) [reply dev/devicelab/lib/framework/devices.dart]: Reverted back to previous formatting to simplify the code review

File: dev/devicelab/lib/framework/devices.dart (hunk 6 above)
Comment (godofredoc) [reply dev/devicelab/lib/framework/devices.dart]: Reverted back to previous formatting to simplify the code review

File: dev/devicelab/lib/framework/devices.dart [hunk 10]
```diff
@@ -636,21 +633,19 @@ class AndroidDevice extends Device {
...
+            stream.sink.add(line);
+          }
+        }, onDone: () {
+          stdoutDone.complete();
+        });
+        process.stderr.transform<String>(utf8.decoder).transform<String>(const LineSplitter()).listen((String line) {
+          print('adb logcat stderr: $line');
+        }, onDone: () {
+          stderrDone.complete();
+        });
```
[the remark this answers is not in the document]
Comment (godofredoc) [reply dev/devicelab/lib/framework/devices.dart]: Reverted back to previous formatting to simplify the code review

File: dev/devicelab/lib/framework/devices.dart (hunk 8 above)
Comment (godofredoc) [reply dev/devicelab/lib/framework/devices.dart]: Reverted back to previous formatting to simplify the code review

File: dev/devicelab/lib/tasks/perf_tests.dart (hunk 9 above)
Comment (godofredoc) [reply dev/devicelab/lib/tasks/perf_tests.dart]: Reverted back to previous formatting to simplify the code review

**Review by dnfield — APPROVED** [review]
LGTM

**Pull request merged**